### PR TITLE
Cody/avro consumers

### PR DESCRIPTION
**Pull request by CJPoll**
Get consumption of avro-serialized topics working.

Explanations in comments of files.

QA Steps:
1. Clone, mix.deps.get, mix compile. Reject PR on compile errors.
2. Uncomment pipeline in `application.ex`. Remember to put a comma on the line before.
3. Run `iex -S mix`. Reject PR if crashes.
4. Verify that the contents of your RR users table are successfully pulled off kafka, decoded, and printed. If you see a bunch of kafka message structs containing avro data using erlang data structures, it has succeeded.

### CODE REVIEW

File: lib/rivulet/measure/message.ex [hunk 1]
```diff
@@ -1,9 +1,21 @@
...
+  end
+
   def latency(key, microseconds) do
     msg(["key", key, "lat", microseconds])
   end
 
+  def utilization(resource, utilization_class, value) when is_integer(value) do
+    utilization(resource, utilization_class, Integer.to_string(value))
+  end
+
```
Comment (CJPoll) [inline lib/rivulet/measure/message.ex]: Fixed some type errors in this module. The Rivulet.Measure(.*) stuff is a proof of concept of a measurement lib for apps. It'll be a good way to figure out any tooling that needs to be built to get a schema into the registry for devs.

File: mix.exs [hunk 2]
```diff
@@ -8,20 +8,25 @@ defmodule Rivulet.Mixfile do
...
-    [applications: [:logger, :kafka_ex],
-     mod: {Rivulet, []}]
+    [applications: [:mix, :logger, :kafka_ex, :eavro, :httpoison, :gen_stage, :poison, :hackney],
+     mod: {Rivulet.Application, []}]
   end
 
   defp aliases do
-    ["compile": ["compile --warnings-as-errors"]]
+    ["compile": ["compile --warnings-as-errors"],
+     "test": ["test --no-start", "dialyzer --halt-exit-status"]]
```
Comment (CJPoll) [inline mix.exs]: Note that we've added dialyzer. As part of QA, please do `mix test`, which runs tests _and_ runs dialyzer.

File: lib/application.ex [hunk 3]
```diff
@@ -0,0 +1,17 @@
+defmodule Rivulet.Application do
+  use Application
+
+  def start(_, _) do
+    import Supervisor.Spec
+
+    children = [
+      supervisor(Registry, [:unique, Rivulet.Registry]),
+      worker(Rivulet.Avro.Cache, [])
+      #worker(Rivulet.TestPipeline, ["raw.postgres.review_rocket.public.users", 0, 0])
```
Comment (brianbroderick) [inline lib/application.ex]: I'd put a comment saying it's for testing.

File: lib/rivulet/avro/registry.ex [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      |> :eavro.parse_schema
+
+    {:ok, schema}
+  catch
+    :exit, {:badarg, nil} -> {:error, :no_schema_provided}
+  end
+
+  @spec get_schema(Avro.schema_id) :: {:ok, Avro.schema} | {:error, reason}
+  def get_schema(schema_id) do
+    with {:ok, %HTTPoison.Response{body: {:ok, schema}}} <- get("/schemas/ids/#{schema_id}") do
```
Comment (aweagel) [inline lib/rivulet/avro/registry.ex]: oob: We talked about pulling in Cody's http mocking stuff so we can test this function (and others like it) as well.

File: lib/rivulet/event_printer.ex [hunk 5]
```diff
@@ -1,29 +1,31 @@
...
+  def init(_) do
     state = %State{}
 
-    {:consumer, state, subscribe_to: [{{:via, Registry, {Rivulet.Registry, "Kafka.#{topic}.#{inspect partition}"}}, max_demand: 10}]}
+    {:consumer, state}
   end
 
   def handle_events(events, _from, %State{count: 100}) do
-    IO.inspect("Batch of 100 completed")
+    Logger.debug("Batch of 100 completed")
```
Comment (aweagel) [inline lib/rivulet/event_printer.ex]: Consider generalizing the `count` here

File: mix.exs [hunk 6]
```diff
@@ -8,20 +8,25 @@ defmodule Rivulet.Mixfile do
...
   end
 
   defp deps do
     [{:kafka_ex, "~> 0.6.5"},
+     {:dialyxir, "~> 0.5.0", only: [:dev, :test], runtime: false},
+     {:eavro, git: "git@github.com:podium/eavro", ref: "fix-complex-primitives", manager: :rebar},
```
Comment (aweagel) [inline mix.exs]: Did you end up doing a PR back onto the main repo for this one?

File: lib/mix/tasks/schema.create.ex [hunk 7]
```diff
@@ -0,0 +1,33 @@
+defmodule Mix.Tasks.Schema.Create do
+  use Mix.Task
```
Comment (CJPoll) [inline lib/mix/tasks/schema.create.ex]: This is a mix task to post the latest versions of avro schemas for a particular topic.

`priv/avro_schemas/#{topic}/key.avsc`
`priv/avro_schemas/#{topic}/value.avsc`

File: lib/rivulet.ex [hunk 8]
```diff
@@ -1,16 +0,0 @@
-defmodule Rivulet do
```
Comment (CJPoll) [inline lib/rivulet.ex]: Moved this module to `lib/rivulet/application.ex`

File: lib/rivulet/avro/cache.ex [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      [] -> nil
+      [{^schema_id, schema}] -> schema # Since this table is a set, there will only ever be 0 or 1 matches.
+    end
+  rescue
+    ArgumentError -> :no_cache
+  end
+
+  # Callback Functions
+  def init([name]) do
+    tid = :ets.new(name, [:set, :public, :named_table, {:keypos, 1}, {:read_concurrency, true}])
```
Comment (CJPoll) [inline lib/rivulet/avro/cache.ex]: Confirmed 👍

File: lib/rivulet/avro/registry.ex [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+      {:ok, %Schema{schema_id: schema_id, schema: schema}}
+    else
+      {:ok, %HTTPoison.Response{body: {:error, _reason} = err}} -> err
+      {:error, _} = err -> err
+    end
+  catch
+    :exit, {:badarg, nil} -> {:error, :no_schema_provided}
+  end
+
+  @spec schema_for(Partition.topic) :: %{key: Schema.t, value: Schema.t} | {:error, term}
```
Comment (CJPoll) [inline lib/rivulet/avro/registry.ex]: TODO: Enumerate the many, many possible values for `{:error, term}`

File: mix.exs (hunk 6 above)
Comment (aweagel) [inline mix.exs]: oob: Not yet. But we should :)

File: mix.exs (hunk 2 above)
Comment (aweagel) [inline mix.exs]: Still not _entirely_ sure this shouldn't be named something other than `test` so tests and static analysis stuff can be run independently

File: lib/mix/tasks/schema.create.ex [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+  def run([topic]) do
+    Application.ensure_all_started(:httpoison)
+    Application.ensure_all_started(:logger)
+
+    schemas_dir = Path.join([File.cwd!(), "priv", "avro_schemas"])
+    key_schema = File.read!(Path.join([schemas_dir, topic, "key.avsc"]))
+    value_schema = File.read!(Path.join([schemas_dir, topic, "value.avsc"]))
+    IO.puts(value_schema)
+    IO.puts(%{schema: value_schema} |> Poison.encode!)
```
Comment (aweagel) [inline lib/mix/tasks/schema.create.ex]: oob: we decided to pull this into `create_schema` below

File: lib/rivulet/kafka/consumer.ex [hunk 12]
```diff
@@ -14,20 +14,15 @@ defmodule Rivulet.Kafka.Consumer do
   when is_binary(topic)
   and topic != ""
   and is_integer(partition) do
-    GenStage.start_link(__MODULE__, {topic, partition},
-                        name: registry_name(topic, partition))
+    GenStage.start_link(__MODULE__, {topic, partition})
```
Comment (CJPoll) [inline lib/rivulet/kafka/consumer.ex]: Should pass GenStage.start_link opts through start_link

**Pull request merged**